hxremove.1
.de d \" begin display
.sp
.in +4
.nf
..
.de e \" end display
.in -4
.fi
.sp
..
.TH "HXREMOVE" "1" "10 Jul 2011" "7.x" "HTML-XML-utils"
.SH NAME
hxremove \- remove elements from an XML file by means of a CSS selector
.SH SYNOPSIS
.B hxremove
.RB "[\| " \-i " \|]"
.RB "[\| " \-l
.IR language " \|]"
.I selector
.SH DESCRIPTION
.B hxremove
reads a well-formed XML document from standard input and writes it to
standard output without any elements that match the CSS selector that
is given as argument. For example
.d
.B hxremove ol li:first-child
.e
removes the first li (list item in XHTML) from every ol (ordered
list).
.PP
Assumes that class selectors (".foo") refer to an attribute called
"class". And assumes that ID selectors ("#foo") refer to an attribute
called "id".
.PP
To handle HTML files, make them well-formed XML first, e.g., with
.BR "hxnormalize -x" .
.PP
Compare with
.BR hxselect ,
which removes everything
.I but
the selected elements.
.SH OPTIONS
The following options are supported:
.TP 10
.B \-i
Match case-insensitively. Useful for HTML and some other
SGML-based languages.
.TP
.BI \-l " language"
Sets the default language, in case the root element doesn't
have an xml:lang attribute (default: none). Example:
.B \-l en
.TP
.B \-?
Show command usage.
.SH OPERANDS
The following operand is supported:
.TP
.I selector
A selector. Most selectors from CSS level 3 are supported, with the
exception of pseudo-classes, pseudo-elements and selectors with
\fBlast\-\fR in their name.
.SH "SEE ALSO"
.BR asc2xml (1),
.BR xml2asc (1),
.BR hxnormalize (1),
.BR hxselect (1),
.BR UTF-8 " (RFC 2279)"


ChangeLog
2016-10-17  Bert Bos  <bert@w3.org>

	* textwrap.c, hxnormalize.c, normalize7.sh, normalize8.sh: Allow a
	line break before the final ">" of a tag. Line length (option -l)
	is now counted in characters, not bytes, at least if is UTF-8.
	Added two tests for these new features.

2016-05-10  Bert Bos  <bert@w3.org>

	* hxunpipe.c: Added option -b to escape SGML delimiters.

2016-04-14  Bert Bos  <bert@w3.org>

	* Released version 7.0.

	* hxwls.c: Added option -a to output URLs in ASCII (using punycode
	and %-escaping if needed). Converting Internationalized Domain
	Names to ASCII (with punycode) requires libidn or libidn2.

2016-04-08  Bert Bos  <bert@w3.org>

	* wls3.sh, wls5.sh: Tests for hxwls (entity expansion, ASCII
	conversion)

	* hxwls.c: Expand entities in URLs, into ASCII characters if
	possible, or into %-escaped UTF-8 octets otherwise.

2016-03-10  Bert Bos  <bert@w3.org>

	* tree.c (main): dumptree() now takes a FILE argument.

2016-03-07  Bert Bos  <bert@w3.org>

	* hxnormalize.c: Attempt at inserting <![CDATA[...]]> into <style>
	and <script> when going from HTML to XML and removing it again
	when going from XML to HTML. (Problem: We know whether the output
	is XML or not, but not whether the input is supposed to be XML...)

2016-03-04  Bert Bos  <bert@w3.org>

	* hxindex.c: Added options -O and -X to restrict the elements to
	index.

2015-07-21  Bert Bos  <bert@w3.org>

	* tree.c: When a CDATA element (i.e., <script> or <style>) is
	opened, instruct the lexical scanner to only recognize the
	element's end tag and treat everything else as text. (Note that
	this matches HTML5, but not SGML: in SGML, any "</foo" or even
	"</>" would end the CDATA element.)

	* normalize3.sh: new test to check parsing of CDATA elements.

	* normalize4.sh: new test to check parsing of CDATA elements.

	* scan.l: Added an initial condition CDATA that can be started
	from the tree builder (tree.c) when a CDATA element is opened.

	* dtd.hash: Added a field to indicate that an element is defined
	as CDATA

2015-03-12  Bert Bos  <bert@w3.org>

	* Makefile.am: Added another test: normalize2.sh. Added header2.c
	to the sources for hxprintlinks. Use groff instead of man2html to
	make HTML versions of the man pages.

	* dtd.hash: RDFA, in contrast to HTML, says that <meta> may be
	used as phrase-level element as well as in the <head>. The HTML
	tree now allows meta as RDFA says, at the cost of less error
	correction.

2015-01-23  Bert Bos  <bert@w3.org>

	* hxprintlinks: Replaced the Perl version by a (better) C
	version. Added corresponding tests.

2015-01-22  Bert Bos  <bert@w3.org>

	* hxtoc.c: Added option -f (flatten) to remove element structure
	from headers when they are copied to the ToC.

2014-10-21  Bert Bos  <bert@w3.org>

	* Released version 6.9.

	* scan.l: Don't declare strndup() as a function if it is in fact a
	macro. (If _GNU_SOURCE is defined and the compiler is called with
	-O, strndup() may be a macro instead of a function.)

2014-10-06  Bert Bos  <bert@w3.org>

	* hxtoc.c: HTML5 now has a "header" element instead of an
	"hgroup."

	* dtd.hash: Added HTML5 elements header, aside, section, article,
	nav and figure. They currently behave very much like a div. Added
	figcaption, which is allowed inside figure.

2014-10-02  Bert Bos  <bert@w3.org>

	* extract1.sh: Explicitly use 127.0.0.1 (an IPv4 address) instead
	of localhost, because the latter may resolve to ::1 (an IPv6
	address), which recent libcurl will use, but nc by default only
	listens to IPV4 traffic.

2014-09-02  Bert Bos  <bert@w3.org>

	* hxnormalize.c: A system literal is only required in the DOCTYPE
	of an XML document if there also is a public literal.

2014-08-18  Bert Bos  <bert@w3.org>

	* types.c: removed modifier "inline" from min() and max()
	functions so that hxindex compiles also with clang. clang, unlike
	gcc, but in conformance with the C99 standard, considers inline
	functions as static function (i.e., not exported).

	* openurl.c: fixed bug, where wait_for_data() was meant to return
	a code of type CURLcode, but actually returned CURLMcode.

	* Added option -s (rewrite links to self) to hxcopy.

2014-07-21  Bert Bos  <bert@w3.org>

	* Makefile.am: "make install" now also installs the COPYING file
	in docdir (by default /usr/local/share/doc/html-xml-utils/)

2014-04-02  Bert Bos  <bert@w3.org>

	* hxnormalize.c: Option -x now also creates an (empty) system
	literal in the doctype if the input didn't have one.

	* scan.l: Also escape '<' and '>' in attributes. Allowed but not
	needed for SGML (and hence HTML4), but required by XML.

2014-03-14  Bert Bos  <bert@w3.org>

	* tests/extract1.sh: Now trying with and without a -p option to
	nc, as that option is required on some versions of nc and
	disallowed on others. Its absence caused the test to hang on some
	platforms. Also removed bash-specific syntax, for more
	portability.
	(Thanks to Anthony J. Bentley <anthony [at] cathet.us>)

2014-03-12  Bert Bos  <bert@w3.org>

	* Released 6.7

	* openurl.c: Include fopencookie.h instead of fopencookie.e

	* fopencookie.c: Don't use cexport to generate a header file
	fopencookie.e but use a static fopencookie.h. (We need to
	distribute this header file, but the .e file may be empty.)

	* bootstrap.sh: No need for "-I m4" anymore, because the option is
	now in the Makefile.am.

	* Makefile.am: Need to make cexport before .e files can be
	generated. (The fix for distcheck wasn't complete yet.) Don't
	generate fopencookie.e, but provide a static fopencookie.h and
	make most programs depend on that.

2014-03-11  Bert Bos  <bert@w3.org>

	* Released 6.6

	* Makefile.am: (Finally :-) ) fixed an error reported by "make
	distcheck" (files left after distclean), by depending .e files on
	cexport.c, rather than on the binary cexport.

	* Makefile.am: Added "ACLOCAL_AMFLAGS = -I m4" to make aclocal
	search configuration macros in the directory m4.

	* hxcite.c, hxindex.c, hxmkbib.c, genid.c: Removed duplicate

hxremove.c
/*
 * Copy input (well-formed XML) without any elements that match a given selector
 *
 * To do: Put common routines with hxselect in separate module.
 *
 * To do: Only look for a LANG attribute if the input is XHTML.
 *
 * Part of HTML-XML-utils, see:
 * http://www.w3.org/Tools/HTML-XML-utils/
 *
 * Copyright © 2012 World Wide Web Consortium
 * See http://www.w3.org/Consortium/Legal/copyright-software
 *
 * Author: Bert Bos <bert@w3.org>
 * Created: 21 Oct 2012
 */
#include "config.h"
#include <assert.h>
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <ctype.h>
#include <stdbool.h>
#ifdef HAVE_STRING_H
# include <string.h>
#elif HAVE_STRINGS_H
# include <strings.h>
#endif
#include "types.e"
#include "tree.e"
#include "selector.e"
#include "heap.e"
#include "errexit.e"
#include "html.e"
#include "scan.e"


static Selector selector;			/* The selector to match */
static string language = "";			/* Initial language */
static bool case_insensitive = false;	/* How to match elems/attrs */


/* get_language -- return the (inherited) human language of an element */
static conststring get_language(const Node *n)
{
  conststring s;

  assert(n);
  if (n->tp == Root) return language; /* Language from -l option */
  assert(n->tp == Element);
  if ((s = get_attrib(n, "xml:lang"))) return s;
  if ((s = get_attrib(n, "lang"))) return s;
  return get_language(n->parent);
}


/* same -- compare two names, case-(in)sensitively, depending */
static bool same(const string a, const string b)
{
  return case_insensitive ? strcasecmp(a, b) == 0 : eq(a, b);
}


/* nr_sisters -- return # of elder sisters of a node */
static int nr_sisters(const Node *n)
{
  if (!n->sister) return 0;
  else return 1 + nr_sisters(n->sister);
}


/* nr_typed_sisters -- return # of elder sisters of type t */
static int nr_typed_sisters(const Node *n, const Node *t)
{
  if (!n->sister)
    return 0;
  else if (!same(n->sister->name, t->name))
    return nr_typed_sisters(n->sister, t);
  else
    return 1 + nr_typed_sisters(n->sister, t);
}


/* get_attr -- return the value of the named attribute, or NULL */
static string get_attr(const Node *n, const string name)
{
  pairlist p;

  for (p = n->attribs; p && !same(p->name, name); p = p->next) ;
  return p ? p->value : NULL;
}


/* includes -- check for word in the space-separated words of line */
static bool includes(const string line, const string word)
{
  int i = 0, n = strlen(word);

  /* What should happen if word is the empty string? */
  /* To do: compare with contains() in class.c, keep the best */
  while (line[i]) {
    if (case_insensitive) {
      if (!strncasecmp(line+i, word, n) && (!line[i+n] || isspace(line[i+n])))
	return true;
    } else {
      if (!strncmp(line+i, word, n) && (!line[i+n] || isspace(line[i+n])))
	return true;
    }
    do i++; while (line[i] && !isspace(line[i]));
    while (isspace(line[i])) i++;
  }
  return false;
}


/* starts_with -- check if line starts with prefix */
static bool starts_with(const string line, const string prefix)
{
  return case_insensitive
    ? strncasecmp(line, prefix, strlen(prefix)) == 0
    : strncmp(line, prefix, strlen(prefix)) == 0;
}


/* ends_with -- check if line ends with suffix */
static bool ends_with(const string line, const string suffix)
{
  int n1 = strlen(line), n2 = strlen(suffix);
  return n1 >= n2 && eq(line + n1 - n2, suffix);
}


/* contains -- check if line contains s */
static bool contains(const string line, const string s)
{
  return strstr(line, s) != NULL;
}


/* lang_match -- check if language specific is subset of general */
static bool lang_match(const conststring specific, const conststring general)
{
  int n = strlen(general);
  return !strncasecmp(specific, general, n)
    && (specific[n] == '-' || !specific[n]);
}


/* simple_match -- check if a node matches a simple selector */
static bool simple_match(const Node *n, const SimpleSelector *s)
{
  AttribCond *p;
  PseudoCond *q;
  string h;
  int i;

  /* Match the type selector */
  if (s->name && !same(s->name, n->name)) return false;

  /* Match the attribute selectors, including class and ID */
  for (p = s->attribs; p; p = p->next) {
    if (!(h = get_attr(n, (p->op == HasClass) ? (string)"class"
		       : (p->op == HasID) ? (string) "id" : p->name)))
      return false;
    switch (p->op) {
    case Exists: break;
    case Equals:
    case HasID: if (!eq(p->value, h)) return false; break;
    case Includes:
    case HasClass: if (!includes(h, p->value)) return false; break;
    case StartsWith: if (!starts_with(h, p->value)) return false; break;
    case EndsWidth: if (!ends_with(h, p->value)) return false; break;
    case Contains: if (!contains(h, p->value)) return false; break;
    case LangMatch: if (!lang_match(h, p->value)) return false; break;
    default: assert(!"Cannot happen");
    }
  }

  /* Match the pseudo-classes */
  for (q = s->pseudos; q; q = q->next) {
    switch (q->type) {
    case RootSel:
      if (n->parent) return false;
      break;
    case NthChild:
      i = nr_sisters(n) + 1;
      if (q->a == 0) {if (i != q->b) return false;}
      else {if ((i-q->b)/q->a < 0 || (i-q->b)%q->a != 0) return false;}
      break;
    case NthOfType:
      i = nr_typed_sisters(n, n) + 1;
      if (q->a == 0) {if (i != q->b) return false;}
      else {if ((i-q->b)/q->a < 0 || (i-q->b)%q->a != 0) return false;}
      break;
    case FirstChild:
      if (n->sister) return false;
      break;
    case FirstOfType:
      if (nr_typed_sisters(n, n) != 0) return false;
      break;
    case Lang:
      if (!lang_match(get_language(n), q->s)) return false;
      break;
    default:
      assert(!"Cannot happen");
    }
  }
  return true;
}


/* matches -- check if an element t matches the selector s */
static bool matches(const Tree t, const Selector s)
{
  Tree h;

  assert(s);
  if (! t) return false;
  if (! simple_match(t, s)) return false;
  if (! s->context) return true;
  switch (s->combinator) {
  case Descendant:
    for (h = t->parent; h && !matches(h, s->context); h = h->parent);
    return h != NULL;
  case Child:
    return matches(t->parent, s->context);
  case Adjacent:
    return matches(t->sister, s->context);
  case Sibling:
    for (h = t->sister; h && !matches(h, s->context); h = h->sister);
    return h != NULL;
  default:
    assert(!"Cannot happen");
    return false;
  }
}


/* print_tree -- print the tree below t, omitting elements that match the selector */
static void print_tree(Tree t)
{
  pairlist p;

  if (!t) return;
  switch (t->tp) {
  case Element:
    if (!matches(t, selector)) {
      printf("<%s", t->name);
      for (p = t->attribs; p; p = p->next) printf(" %s=\"%s\"", p->name, p->value);
      if (!t->children) {
	printf("/>");
      } else {
	printf(">");
	print_tree(t->children);
	printf("</%s>", t->name);
      }
    }
    print_tree(t->sister);
    break;
  case Text:
    printf("%s", t->text);
    print_tree(t->sister);
    break;
  case Comment:
    printf("<!--%s-->", t->text);
    print_tree(t->sister);
    break;
  case Declaration:
    printf("<!DOCTYPE %s", t->name);
    if (t->text && t->url) printf(" PUBLIC \"%s\" \"%s\"", t->text, t->url);
    else if (t->text) printf(" PUBLIC \"%s\"", t->text);
    else if (t->url) printf(" SYSTEM \"%s\"", t->url);
    printf(">");
    print_tree(t->sister);
    break;
  case Procins:
    printf("<?%s>", t->text);
    print_tree(t->sister);
    break;
  case Root:
    print_tree(t->children);

    break;
  default: assert(!"Cannot happen!");
  }
}


/*********************** Parser callback API ***********************/

/* handle_error -- called when a parse error occurred */
void handle_error(void *clientdata, const string s, int lineno)
{
  fprintf(stderr, "%d: %s\n", lineno, s);
}


/* start -- called before the first event is reported */
void* handle_start(void)
{
  Tree *tp;
  new(tp);
  *tp = create();		/* Create an empty tree */
  return tp;
}


/* end -- called after the last event is reported */
void handle_end(void *clientdata)
{
  Tree *t = (Tree*)clientdata;
  print_tree(get_root(*t)); /* Print tree, filtering out unwanted elements */
  tree_delete(*t);
}


/* handle_comment -- called after a comment is parsed */
void handle_comment(void *clientdata, const string commenttext)
{
  Tree *t = (Tree*)clientdata;
  *t = append_comment(*t, commenttext);
}


/* handle_text -- called after a text chunk is parsed */
void handle_text(void *clientdata, const string text)
{
  Tree *t = (Tree*)clientdata;
  *t = tree_append_text(*t, text);
}


/* handle_decl -- called after a declaration is parsed */
void handle_decl(void *clientdata, const string gi, const string fpi, const string url)
{
  Tree *t = (Tree*)clientdata;
  *t = append_declaration(*t, gi, fpi, url);
}


/* handle_pi -- called after a Processing Instruction is parsed */
void handle_pi(void *clientdata, const string pi_text)
{
  Tree *t = (Tree*)clientdata;
  *t = append_procins(*t, pi_text);
}


/* handle_starttag -- called after a start tag is parsed */
void handle_starttag(void *clientdata, const string name, pairlist attribs)
{
  Tree *t = (Tree*)clientdata;
  *t = tree_push(*t, name, attribs);
}


/* handle_emptytag -- called after an empty tag is parsed */
void handle_emptytag(void *clientdata, const string name, pairlist attribs)
{
  Tree *t = (Tree*)clientdata;
  *t = tree_push(*t, name, attribs);
  *t = tree_pop(*t, name);
}


/* handle_endtag -- called after an endtag is parsed (name may be "") */
void handle_endtag(void *clientdata, const string name)
{
  Tree *t = (Tree*)clientdata;
  *t = tree_pop(*t, name);
}


/* handle_endincl -- called at the end of an included file */
/* void handle_endincl(void *clientdata) */

/********************* End of parser callbacks *********************/


/* usage -- print usage message and exit */
static void usage(const conststring progname)
{
  errexit("Usage: %s [-v] [-l language] [-i] selector\n", progname);
}


int main(int argc, char *argv[])
{
  string s;
  int c;

  /* Command line options */
  while ((c = getopt(argc, argv, "il:v")) != -1) {
    switch (c) {
    case 'l': language = optarg; break;
    case 'i': case_insensitive = true;  break;
    case 'v': printf("Version: %s %s\n", PACKAGE, VERSION); return 0;
    case '?': usage(argv[0]); break;
    default: assert(!"Cannot happen");
    }
  }

  /* Bind the parser callback routines to our handlers */
  set_error_handler(handle_error);
  set_start_handler(handle_start);
  set_end_handler(handle_end);
  set_comment_handler(handle_comment);
  set_text_handler(handle_text);
  set_decl_handler(handle_decl);
  set_pi_handler(handle_pi);
  set_starttag_handler(handle_starttag);
  set_emptytag_handler(handle_emptytag);
  set_endtag_handler(handle_endtag);

  /* Parse the selector */
  if (optind >= argc) usage(argv[0]);		/* Need at least 1 arg */
  for (s = newstring(argv[optind++]); optind < argc; optind++)
    strapp(&s, " ", argv[optind], NULL);
  selector = parse_selector(s);

  /* Parse the input, build a tree, filter the tree */
  yyin = stdin;
  if (yyparse() != 0) exit(3);
  return 0;
}
